Asher Fredman and Shahar Schnapp

In [4]:
import numpy as np
from sklearn.datasets import make_sparse_spd_matrix
import matplotlib.pyplot as plt
from IPython.display import Math

### 1. Equality constrained optimization
In this section we will find the maximal surface area of a box given the sum of its edges length. <br>
The optimization problem is given by: <br>
$max_{x \in R^3}\{x_1x_2 +x_2x_3 +x_1x_3\}$  $s.t.$  $x_1+x_2+x_3=3 $.

#### (a) Find a critical point for the problem using the Lagrange multiplier method.

Answer:

$L(x, \lambda) = f(x) + \lambda c^{eq}(x)
= x_1x_2 +x_2x_3 +x_1x_3 + \lambda (x_1+x_2+x_3-3)$.

We require that $\nabla L(x, \lambda) =  0$,
meaning $\nabla L(x, \lambda) = 
\begin{bmatrix}
x_2 + x_3 + λ \\
x_1 + x_3 + λ \\
x_1 + x_2 + λ \\
x_1 + x_2 +x_3 -3 
\end{bmatrix} = 
\begin{bmatrix}
0 \\
0 \\
0 \\
0 
\end{bmatrix}$.

We now have 4 equations with 4 unnowns, for which the solution is:

In [198]:
A = np.array([[0,1,1,1],[1,0,1,1],[1,1,0,1],[1,1,1,0]])
b = np.array([[0],[0],[0],[3]])
print(np.linalg.solve(A, b))

[[ 1.]
 [ 1.]
 [ 1.]
 [-2.]]


So we get $x_1=x_2=x_3=1$, and  $λ=-2$.

#### (b) Show that this critical point is a maximum point.

Answer:

$\nabla_x^2L = 
\begin{bmatrix}
0 & 1 & 1 \\
1 & 0 & 1 \\
1 & 1 & 0 
\end{bmatrix}$

In order for the critical point to be a maximum point, we need to show that $y^\top \nabla_x^2L y < 0$ for vectors $y \neq 0$ who satisfy $y^⊤1 = y_1 + y_2 + y_3 = 0$.<br>

$y^\top \nabla_x^2L y = 
\begin{bmatrix}
y_2+y_3 & y_1+y_3 & y_1+y_2 
\end{bmatrix}
\begin{bmatrix}
y_1 \\
y_2 \\
y_3 
\end{bmatrix}$.

Since $y^⊤1 = y_1 + y_2 + y_3 = 0$, then we have that:<br>
$y_2 + y_3 = -y_1$<br>
$y_1 + y_3 = -y_2$<br>
$y_1 + y_2 = -y_3$

$\Rightarrow y^\top \nabla_x^2L y = 
\begin{bmatrix}
-y_1 & -y_2 & -y_3 
\end{bmatrix}
\begin{bmatrix}
y_1 \\
y_2 \\
y_3 
\end{bmatrix} = -(y_1^2 + y_2^2 + y_3^2) < 0$ for any $y \neq 0$.

### 2. General constrained optimization
Assume that we have the following problem:<br><br>
$min_{x \in R^2}\{(x_1+x_2)^2 -10(x_1+x_2)\}  
s.t.  \begin{cases}
  3x_1 + x_2 = 6\\    
  x_1^2 + x_2^2 \leq 5\\  
  -x_1 \leq 0
\end{cases}$

#### (a) Find a critical point $x^∗$ using the Lagrange multipliers method for which the fewest inequality constraints are active. Show that the KKT conditions hold. (no need to determine whether this point is a minimum or maximum).

Answer:

Let us define:<br><br>
$x:=[x_1, x_2]^\top$<br><br>
$f(x) = (x_1+x_2)^2 -10(x_1+x_2)$<br><br>
$c^{eq} := [3x_1 + x_2 - 6 = 0]$<br><br>
$c^{ieq} := 
\begin{bmatrix}
x_1^2 + x_2^2 -5 \leq 0 \\
-x_1 \leq 0
\end{bmatrix}$<br><br>
Where as we learned, the lagrangian for this case is:<br><br>
$L(x, λ^{eq} , λ^{ieq}) = f(x) + (λ^{eq})^⊤ c^{eq}(x) + (λ^{ieq})^⊤c^{ieq}(x)$

For the KKT conditions to hold we need to show that for a given local solution $x^*$ the following conditions are satisfied:<br>
1. $\nabla_x L(x^*, λ^{eq^*} , λ^{ieq^*}) = 0$<br>
2. $c^{eq}(x^*) = 0$<br>
3. $c^{ieq}(x^*) \leq 0$<br>
4. $λ^{ieq^*} \geq 0$<br>
5. $λ_l^{ieq^*}c^{ieq}(x^*) = 0$<br>

We will now go over all the combinations of possibly active $c^{ieq}$'s which satisy the KKT conditions, and we will start with the smallest combination possible.<br>
As we learned for the active $c^{ieq}$'s condition for is $λ^{ieq^*} > 0$ and for the non-active ones it's $λ^{ieq^*} = 0$ (so we can ignore the non-active c's).

##### option 1: there are no active $c^{ieq}$'s.
In that case:<br><br>
$L(x, λ^{eq} , λ^{ieq})$<br> 
$=f(x) + (λ^{eq})^⊤ c^{eq}(x) + (0)^⊤c^{ieq}(x)$<br>
$=(x_1+x_2)^2 -10(x_1+x_2) + (λ_1^{eq})^⊤(3x_1 + x_2 - 6) $

$\nabla L(x, \lambda) = 
\begin{bmatrix}
2x_1 + 2x_2 -10 + 3λ_1 \\
2x_1 + 2x_2 -10 + λ_1 \\
3x_1 + x_2 -6
\end{bmatrix} = 
\begin{bmatrix}
0 \\
0 \\
0 
\end{bmatrix}$.

We now have 3 equations with 3 unnowns, for which the solution is:

In [197]:
A = np.array([[2,2,3],[2,2,1],[3,1,0]])
b = np.array([[10],[10],[6]])
print(np.linalg.solve(A, b))

[[0.5]
 [4.5]
 [0. ]]


So we get $x_1=0.5, x_2=4.5$, and  $λ_1=0$.
Placing these values into $c^{eq}$ we see that the equality does not hold, and therefor the KKT conditions are not satisfied for this combination:<br>
$c^{eq}(x) = 1.5-4.5-6 = -9 \neq 0$.

##### option 2: there is one active $c^{ieq}$.
Let's assume that only the first $c^{ieq}$ is active (and so $λ^{ieq} = [λ_2,0]^\top$):<br><br>
$L(x, λ^{eq} , λ^{ieq})=f(x) + (λ^{eq})^⊤ c^{eq}(x) + (0)^⊤c^{ieq}(x)$<br>
$=(x_1+x_2)^2 -10(x_1+x_2) + (λ_1^{eq})^⊤(3x_1 + x_2 - 6) + (λ_2^{ieq})^⊤(x_1^2 + x_2^2 -5)$

$\nabla L(x, \lambda) = 
\begin{bmatrix}
2x_1 + 2x_2 -10 + 3λ_1 + 2λ_2x_1\\
2x_1 + 2x_2 -10 + λ_1 + 2λ_2x_2\\
3x_1 + x_2 -6 \\
x_1^2 + x_2^2 -5
\end{bmatrix}$

We'll first find $x_1$ and $x_2$ using the last two equations.<br><br>
$x_2 = 6-3x_1$<br>
$x_1^2 + (6-3x_1)^2 = 5$<br>
$x_1^2 + 9x_1^2 + 36 -36x_1 = 5$<br>
$10x_1^2 - 36x_1 +31 = 0$<br><br>
Solving the above quedratic equation we get two possible solutions:<br>
$x_{1_a,2_a} = [2.174, -0.522]$<br>
$x_{1_b,2_b} = [1.426, 1.722]$<br>


It can easily be validated that both solutions satisfy the $c^{eq}$ constrained. Now we move on the find the λ's.<br><br>
Using the first two equations with $x_{1_a,2_a}$ we get:<br>
1. $λ_{2_a} = 
\frac{5-x_2}{x_1} - 1 - \frac{1.5λ_{1_a}}{x_1} = 
1.54-0.6λ_{1_a}$<br>
2. $λ_{2_a} = 
\frac{5-x_1}{x_2} - 1 - \frac{0.5λ_{1_a}}{x_2} = 
-6.41+0.9λ_{1_a}$<br>
$\Rightarrow 1.5λ_{1_a} = 7.95
\Rightarrow λ_{1_a,2_a} = [5.3, -1.64]$

Notice that $λ_{2_a} < 0$ which does not satisfy the 4TH KKT condition, so we move on the check the values of λ using $x_{1_b,2_b}$:<br>
1. $λ_{2_b} = 
\frac{5-x_2}{x_1} - 1 - \frac{1.5λ_{1_b}}{x_1} = 
0.3 - 0.07λ_{1_a}$<br>
2. $λ_{2_b} = 
\frac{5-x_1}{x_2} - 1 - \frac{0.5λ_{1_b}}{x_2} = 
1.135-0.29λ_{1_b}$<br>
$\Rightarrow 0.22λ_{1_b} = 0.835
\Rightarrow λ_{1_b,2_b} = [3.8, 0.035]$

We are left to validate that the 5TH KKT condition holds using the $λ^{ieq}$ and the $x$ we found, and indeed:<br>
$c^{ieq}[2](x) = x_1^2 + x_2^2 -5 =
1.426^2 + 1.722^2 - 5 = 0$<br>
(for the non-active $c^{ieq}$ we take $λ=0$ and it is immediate).

#### (b) Write the unconstrained minimization problem that corresponds to the problem above using the penalty method with $ρ(x) = x^2$ as a penalty function.

Find $min_{x \in R^n}\{f_\mu(x)\}$ where<br> 
$f_\mu(x) := (x_1+x_2)^2 -10(x_1+x_2) + 
\mu\times[(3x_1 + x_2 - 6)^2 + (max\{0, x_1^2 + x_2^2 - 5\})^2 + (max\{0, -x_1\})^2]$

#### (c) Use the penalty method to get the minimizer $x^∗$ up to two digits of accuracy. Solve the optimization problems using steepest descent with Armijo linesearch.

In [1]:
f = lambda x : (x[0]+x[1])**2 -10*(x[0]+x[1])

def get_f_mu(mu):
    return lambda x,f,c : f(x) + mu*((c[0](x))**2 + max(0,c[1](x))**2 + max(0,c[2](x))**2)

def get_constraintes():
    return [lambda x : 3*x[0] + x[1] -6, 
            lambda x : x[0]**2 + x[1]**2 -5, 
            lambda x : -x[0]]

def get_g_mu(mu):
    return [lambda x,c : 2*(x[0]+x[1]) - 10 + 6*mu * (c[0](x)) + mu*2*x[0] * (max(0, c[1](x)) * c[1](x)) -mu * (max(0, c[2](x)) * c[2](x)), 
            lambda x,c : 2*(x[0]+x[1]) - 10 + 2*mu * (c[0](x)) + mu*2*x[1] * (max(0, c[1](x)) * c[1](x))]

In [47]:
def linesearch(f_mu,x,f,constraintes,d,gk,alpha0,beta,c):
    alphaj = alpha0
    for jj in range(20):
        x_temp = np.ndarray.tolist(np.array(x) + alphaj*d)
        if f_mu(x_temp,f,constraintes) <= f_mu(x,f,constraintes) + alphaj*c*np.dot(d,gk):
            break
        else:
            alphaj = alphaj*beta
    return alphaj

In [48]:
mus = [0.01, 0.1, 1, 10, 100]
beta = 0.5
c = 1e-4
alpha_prev = 1.0
x_SD = [-1.4, 2.0]
constraintes = get_constraintes()

for mu in mus:
    f_mu = get_f_mu(mu)
    g_mu = get_g_mu(mu)
    for k in range(500):
        gk = np.array([g_mu[0](x_SD, constraintes), g_mu[1](x_SD, constraintes)])
        d_SD = -gk
        alpha_SD = linesearch(f_mu,x_SD,f,constraintes,d_SD,gk,alpha_prev,beta,c)
        x_SD=np.ndarray.tolist(np.array(x_SD)+alpha_SD*d_SD)
print(np.around(x_SD, decimals=3))

[1.41  1.758]


So when solvin the optimization problem using steepest descent with Armijo linesearch we get:<br>
$x_{1,2} = [1.41, 1.758]$ which is pretty close to the $[1.426, 1.722]$ we got before.

### 3. Box-constrained optimization
In this question we will write the projected coordinate descent method for quadratic box-constrained minimization. Assume the following problem<br><br>
$min_{x \in R^n}\{\frac{1}{2}x^\top Hx - x^\top g\}$  s.t. $a\leq x \leq b$,<br><br>
where $H ∈ R^{n×n}$ is symmetric positive definite, $g ∈ R^n$, and $a < b ∈ R^n$ are the lower and upper bounds on the solution x.

#### (a) Give a closed form solution for the scalar box constrained minimization problem $min_{x \in R}\{\frac{1}{2}hx^2 - gx\}$ s.t. $a\leq x \leq b$, for $a < b, h > 0$. 

Answer:

Given $f(x):=\frac{1}{2}hx^2 - gx$ and $c^{ieq} := 
\begin{bmatrix}
a - x \leq 0 \\
x - b \leq 0
\end{bmatrix}$<br><br>
Let us define the following lagrangian function:<br>
$L(x, \lambda) = 
\frac{1}{2}hx^2 - gx + \lambda_1(a - x) + \lambda_2(x - b)$.

$\nabla_x L(x, \lambda) = 
\begin{bmatrix}
hx - g - λ_1 + λ_2
\end{bmatrix}$.<br><br>
$hx - g - λ_1 + λ_2 = 0 \Rightarrow x = \frac{g + λ_1 - λ_2}{h}$<br><br>
This point is a minumim point since $\nabla_x^2 L(x, \lambda) = h > 0$.<br><br>
We'll define $z:= \frac{g + λ_1 - λ_2}{h}$, which gives us:<br><br>
$x^* = \begin{cases}
  z & \text{if }a\le z\le b\\ \\    
  a & \text{if }z < a\\ \\  
  b & \text{if }z > b\\
\end{cases}$

(Note that in order to find $λ_1$ and $λ_2$ we can slipt the possible values of $x$ into three groups:<br>
(1) $a\le z\le b$. For this case since $x$ is in a valid domain then both $λ_1,λ_2 = 0$ and so $z = \frac{g}{h}$<br>
(2) $z < a$. Since we also know that $a<b$, then in this case ovbiously the second inequality in $c^{ieq}$ ($x - b \leq 0$) is true and therefor $λ_2 = 0$.<br>
We get two equations: $x = \frac{g + λ_1}{h}$ and $x = a$, which means that $λ_1 = a\times h - g$.<br>
Since we know that $z = \frac{g}{h} < a$ then $λ_1 = a\times h - g > \frac{g}{h}\times h - g = 0$ and so $λ_1 \geq 0$.<br>
(3) $b < x$. This case is very similar to the second case and so $λ_2 = -b\times h + g$.<br>
Now since we know that $z = \frac{g}{h} > b$ then as before we would get that $λ_1 \geq 0$).<br>

#### (b) In the projected coordinate descent algorithm we sweep over all the variables $x_i$ one by one, and for each solve the box-constrained minimization problem for the scalar variable $x_i$, given that the rest are known. Show that the minimization for each scalar $x_i$ is given by
$min_{x_i \in R}\{\frac{1}{2}h_{ii}x_i^2 + [(\sum_{j\neq i}h_{ij}x_j) - g_i]x_i\}$ s.t. $a_i\leq x_i \leq b_i$

Given that all $x_j$'s are known, where $j\neq i$, we are just left with minimizing $\frac{1}{2}h_{ii}x_i^2 + [(\sum_{j\neq i}h_{ij}x_j) - g_i]x_i$.

(1) $\frac{1}{2}x^\top Hx = 
\frac{1}{2}\sum_{i=1}^{n}x_i\sum_{j=1}^{n}x_jH_{j} = 
\frac{1}{2}\sum_{i=1}^{n}\sum_{i=1}^{n}x_ix_jH_{ji} =$
$= \frac{1}{2}\sum_{i=1}^{n}(x_i^2H_{ii} + \sum_{i\neq j}x_ix_jH_{ji})$<br><br>
(2) $x^\top g = \sum_{i=1}^{n}x_ig_i$

Given $f(x):= \frac{1}{2}x^\top Hx - x^\top g$ and since H is symetric ($h_{ij} = h_{ji}$) we have that:<br><br>
(1 + 2) $\Rightarrow f(x) = \sum_{i=1}^{n} \frac{1}{2}h_{ii}x_i^2 + [(\sum_{j\neq i}h_{ij}x_j) - g_i]x_i$

#### (c)+(d) Use the previous section to write a program for solving the using a projected coordinate descent algorithm, and solve it for the following parameters:

In [209]:
H = -np.ones((5,5))
for i in range(5):
    H[i,i] = 5
a = np.zeros((5,1))
b = 5*np.ones((5,1))
x_initial = 2.5*np.ones((5,1))
g = np.expand_dims(np.array([18.,6.,-12.,-6.,18.]),1)
print("H:")
print(H)
print("g:")
print(g)
print("a:")
print(a)
print("b:")
print(b)

H:
[[ 5. -1. -1. -1. -1.]
 [-1.  5. -1. -1. -1.]
 [-1. -1.  5. -1. -1.]
 [-1. -1. -1.  5. -1.]
 [-1. -1. -1. -1.  5.]]
g:
[[ 18.]
 [  6.]
 [-12.]
 [ -6.]
 [ 18.]]
a:
[[0.]
 [0.]
 [0.]
 [0.]
 [0.]]
b:
[[5.]
 [5.]
 [5.]
 [5.]
 [5.]]


Though it was not required, In addition to the  the coordinate descent solution (which we will show now), we decided to compare the results with the projected steepest descent algorithm.<br><br>
(1) Coordinate descent solution:

In [233]:
def project (x_ind, bottom, top):
    return min(max(x_ind,bottom),top)

In [251]:
def proj_ind(x, i, H, g):
    mult_sum = 0
    for j in range(H.shape[0]):
        if j != i:
            mult_sum += H[i][j] * x[j][0]
    x_sol =  -(mult_sum - g[i][0]) / H[i][i]
    return project(x_sol, a[i][0], b[i][0])

In [252]:
def proj_cd(H, g, x_initial, maxIter, convergence_criterion):
    x_sol = x_initial
    x_l = x_sol.shape[0]
    for itera in range(maxIter):
        for i in range(x_l):
            x_sol[i] = proj_ind(x_sol, i, H, g)
    return x_sol

In [254]:
maxIter = 100
convergence_criterion = 0.001
learned_x = proj_cd(H, g, x_initial, maxIter, convergence_criterion)

The solution is:

In [255]:
print(np.around(learned_x, decimals=3))

[[5.   ]
 [3.667]
 [0.667]
 [1.667]
 [5.   ]]


(2) Comparing the coordinate descent solution with a projected steepest descent solution.

In [256]:
f_proj = lambda x,H,g : 0.5*(x.T.dot(H)).dot(x) - x.T.dot(g)
g_proj = lambda x,H,g : H.dot(x) - g

def proj_line_search(f,x,H,g,d,gk,alpha0,beta,c):
    alphaj = alpha0
    for jj in range(10):
        x_temp = x + alphaj*d
        curr_val = f_proj(x_temp,H,g)
        prev_val = f_proj(x,H,g)
        res = alphaj*c*(d.T.dot(gk))
        if curr_val <= prev_val + res:
            break
        else:
            alphaj = alphaj*beta
    return alphaj

In [257]:
def proj_sd(f_proj, g_proj, H, g, x_0, maxIter, convergence_criterion):
    x_prev = x_0
    convergence_prev = np.linalg.norm(f_proj(x_prev,H,g))
    for k in range(maxIter):
        g_prev = g_proj(x_prev,H,g)
        d = -g_prev
        alpha = proj_line_search(f,x_prev,H,g,d,g_prev,alpha_prev,beta,c)
        z_k = x_prev - alpha * g_prev
        x_k = np.minimum(np.maximum(a,z_k), b)
        convergence = np.linalg.norm(f_proj(x_k,H,g))
        convergence_factor = convergence / convergence_prev
        if convergence_factor <= convergence_criterion:
            break
        x_prev = x_k
        convergence_prev = convergence
    return x_prev

In [258]:
maxIter = 100
convergence_criterion = 0.001
learned_x = proj_sd(f_proj, g_proj, H, g, x_initial, maxIter, convergence_criterion)

The solution is:

In [259]:
print(np.around(learned_x, decimals=3))

[[5.   ]
 [3.667]
 [0.667]
 [1.667]
 [5.   ]]
